include/deal.II/base/mpi.templates.h

```diff
@@ -186,6 +186,180 @@ namespace Utilities
+        auto my_proc = dealii::Utilities::MPI::this_mpi_process(comm);
+
+        std::vector<unsigned int> send_to(objects_to_send.size());
+        {
+          unsigned int i=0;
+          for (const auto &m: objects_to_send)
+            send_to[i++] = m.first;
+        }
+
+        const auto receive_from =
+          Utilities::MPI::compute_point_to_point_communication_pattern(comm, send_to);
+
+        std::map<unsigned int, T> received_objects;
+
+        // Separate all MPI_Requests, to allow optimal complexity when sending data around
+        std::vector<MPI_Request> size_send_requests(send_to.size());
+        std::vector<MPI_Request> buffer_send_requests(send_to.size());
+
+        std::vector<MPI_Request> size_receive_requests(receive_from.size());
+        std::vector<MPI_Request> buffer_receive_requests(receive_from.size());
+
+        std::vector<std::vector<char> > buffers_to_send(receive_from.size());
+        std::vector<std::vector<char> > buffers_to_receive(send_to.size());
+
+        std::vector<unsigned int> receive_size(receive_from.size(),0);
+
+        {
+          unsigned int i = 0;
+          for (const auto &rank_and_obj : objects_to_send)
+            buffers_to_send[i++] = Utilities::pack(rank_and_obj.second);
+        }
+
+        {
+          unsigned int i = 0;
+          // Sending sizes
+          for (const auto &rank_obj : objects_to_send)
+            {
+              const auto &rank = rank_obj.first;
+              const unsigned int count = buffers_to_send[i].size();
+
+              const int ierr = MPI_Isend(&count, 1, MPI_UNSIGNED,
+                                         rank, n_procs*rank + my_proc, comm, &size_send_requests[i]);
+              AssertThrowMPI(ierr);
+              ++i;
+            }
+        }
+        // Receiving sizes
+        {
+          unsigned int i = 0;
+          for (const auto &rank : receive_from)
+            {
+              const int ierr = MPI_Irecv(&(receive_size[i]), 1, MPI_UNSIGNED,
+                                         rank, n_procs*my_proc + rank, comm, &size_receive_requests[i]);
+              AssertThrowMPI(ierr);
+              ++i;
+            }
+        }
+
+        // Sending buffers
+        {
+          unsigned int i = 0;
+          for (const auto &rank_obj : objects_to_send)
+            {
+              const auto &rank = rank_obj.first;
+              const auto &buffer = buffers_to_send[i];
+              const int ierr = MPI_Isend(buffer.data(), buffer.size(), MPI_CHAR,
+                                         rank, n_procs*rank + my_proc, comm, &buffer_send_requests[i]);
+              AssertThrowMPI(ierr);
+              ++i;
+            }
+        }
+
+        // Receiving buffers
+        {
+          unsigned int i = 0;
+          for (const auto rank : receive_from)
+            {
+              // We need to have the correct size before we can proceed
+              MPI_Wait(&size_receive_requests[i],MPI_STATUSES_IGNORE);
```

Comment:
I think it would make more sense to:
1. not send the size
2. Do an ``MPI_Probe`` here and extract the size of the message from the status
3. do an ``MPI_recv`` instead of ``Irecv``.
4. Immediately unpack each package as it arrives (so skip the loop below)

Reply:
Do you have evidence that what you propose is better? Non-blocking communication, in my experience, tend to be better when you have potentially very unbalanced data to send and receive. Each `MPI_Probe` and `MPI_recv` is blocking. Here blocking communication only happens when strictly necessary. 

I'd rather keep it as is, unless you have evidence that what you suggest is a better approach.

Reply:
I have no evidence/benchmarks, but we are only waiting if no packet have arrived yet (and that would mean you would also wait when you use nonblocking recv). The advantage with my suggestion is that you can already unpack/process data as it comes in.